File: .agent/skills/transport-data/scripts/generate_static_json.py
import csv
import json
import sys
from datetime import datetime, timedelta
from zoneinfo import ZoneInfo

# Use Paris timezone for GTFS schedules
PARIS_TZ = ZoneInfo('Europe/Paris')

# Target line and stop by NAME (more resilient to ID changes)
TARGET_ROUTE_NAMES = ['E1']  # E1 is the new name
TARGET_STOP_NAMES = ['GERZAT Champfleuri', 'Patural']
TARGET_MAIN_STOP = 'GERZAT Champfleuri'

# Generate for today and next 7 days
start_date = datetime.now(PARIS_TZ)
dates = [(start_date + timedelta(days=i)).strftime('%Y%m%d') for i in range(8)]

schedule_data = {}

# Find route_id by name
target_route_ids = set()
with open('gtfs_data/routes.txt', 'r', encoding='utf-8-sig') as f:
    reader = csv.DictReader(f)
    for row in reader:
        if row['route_short_name'] in TARGET_ROUTE_NAMES:
            target_route_ids.add(row['route_id'])
            print(f"✅ Found route: {row['route_short_name']} -> route_id={row['route_id']}")

if not target_route_ids:
    print(f"❌ Error: Could not find route with names {TARGET_ROUTE_NAMES}")
    sys.exit(1)

# Export route config for TypeScript consumption
gtfs_config = {
    'routeIds': list(target_route_ids),
    'routeNames': TARGET_ROUTE_NAMES,
    'generatedAt': datetime.now(PARIS_TZ).isoformat()
}
with open('src/data/gtfs_config.json', 'w') as f:
    json.dump(gtfs_config, f, indent=2)
print(f"✅ Exported gtfs_config.json with route IDs: {list(target_route_ids)}")

# Find stop_id by name
target_stop_ids = set()
main_stop_ids = set() # specifically for Gerzat Champfleuri
patural_stop_ids = set() # specifically for Patural

with open('gtfs_data/stops.txt', 'r', encoding='utf-8-sig') as f:
    reader = csv.DictReader(f)
    for row in reader:
        for target_name in TARGET_STOP_NAMES:
            if target_name.lower() in row['stop_name'].lower():
                target_stop_ids.add(row['stop_id'])
                if TARGET_MAIN_STOP.lower() in row['stop_name'].lower():
                    main_stop_ids.add(row['stop_id'])
                if 'patural' in row['stop_name'].lower():
                    patural_stop_ids.add(row['stop_id'])
                print(f"✅ Found stop: {row['stop_name']} -> stop_id={row['stop_id']}")

if not target_stop_ids:
    print(f"❌ Error: Could not find any target stops")
    sys.exit(1)

# Update config with stop IDs
gtfs_config['stopIds'] = {
    'all': list(target_stop_ids),
    'champfleuri': list(main_stop_ids),
    'patural': list(patural_stop_ids)
}
with open('src/data/gtfs_config.json', 'w') as f:
    json.dump(gtfs_config, f, indent=2)
print(f"✅ Updated gtfs_config.json with Stop IDs")

# Load calendar
services = {}
with open('gtfs_data/calendar.txt', 'r', encoding='utf-8-sig') as f:
    reader = csv.DictReader(f)
    for row in reader:
        services[row['service_id']] = row

# Load calendar_dates
exceptions = {}
try:
    with open('gtfs_data/calendar_dates.txt', 'r', encoding='utf-8-sig') as f:
        reader = csv.DictReader(f)
        for row in reader:
            if row['date'] not in exceptions:
                exceptions[row['date']] = []
            exceptions[row['date']].append(row)
except FileNotFoundError:
    pass

# Load trips for target routes
trips = {}
with open('gtfs_data/trips.txt', 'r', encoding='utf-8-sig') as f:
    reader = csv.DictReader(f)
    for row in reader:
        if row['route_id'] in target_route_ids:
            trips[row['trip_id']] = row

print(f"📋 Found {len(trips)} trips for target routes")

# Load stop times for target stops
trip_stops = []
with open('gtfs_data/stop_times.txt', 'r', encoding='utf-8-sig') as f:
    reader = csv.DictReader(f)
    for row in reader:
        if row['trip_id'] in trips and row['stop_id'] in target_stop_ids:
            trip_stops.append(row)

print(f"📋 Found {len(trip_stops)} stop times at target stops")

# Build schedule
final_schedule = []

for date_str in dates:
    dt = datetime.strptime(date_str, '%Y%m%d').replace(tzinfo=PARIS_TZ)
    day_of_week = dt.weekday() # 0=Mon, 6=Sun
    days_key = ['monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday', 'sunday'][day_of_week]
    
    active_services = set()
    
    # Check calendar
    for service_id, service in services.items():
        if service['start_date'] <= date_str <= service['end_date']:
            if service[days_key] == '1':
                active_services.add(service_id)
    
    # Check exceptions
    if date_str in exceptions:
        for exc in exceptions[date_str]:
            if exc['exception_type'] == '1':
                active_services.add(exc['service_id'])
            elif exc['exception_type'] == '2':
                active_services.discard(exc['service_id'])
    
    # Filter trips
    # We might have multiple stop_times for the same trip (if it stops at both Patural and Champfleuri)
    # We want to prefer Champfleuri if present, otherwise Patural.
    
    trip_events = {} # trip_id -> {stop_id, arrival, departure}
    
    for stop_time in trip_stops:
        trip_id = stop_time['trip_id']
        stop_id = stop_time['stop_id']
        
        # If we already have an event for this trip, check if we should overwrite it
        if trip_id in trip_events:
            existing_stop = trip_events[trip_id]['stop_id']
            # If current is main stop (Champfleuri) and existing is NOT, overwrite
            if stop_id in main_stop_ids and existing_stop not in main_stop_ids:
                 trip_events[trip_id] = stop_time
            # Else keep existing
        else:
            # STRICT RULE: For Patural (not main stop), ONLY keep Express trips
            # Express Outbound: Destination 'Ballainvilliers'
            # Express Inbound: Destination 'Patural' (or 'Le Patural')
            if stop_id not in main_stop_ids:
                trip_headsign = trips[trip_id]['trip_headsign'].upper()
                if 'BALLAINVILLIERS' not in trip_headsign and 'PATURAL' not in trip_headsign:
                    continue 

            trip_events[trip_id] = stop_time

    for trip_id, stop_time in trip_events.items():
        trip = trips[trip_id]
        if trip['service_id'] in active_services:
            # Convert time to timestamp
            # Handle times > 24:00:00
            h, m, s = map(int, stop_time['arrival_time'].split(':'))
            trip_dt = dt + timedelta(hours=h, minutes=m, seconds=s)
            timestamp = int(trip_dt.timestamp())
            
            final_schedule.append({
                'tripId': trip_id,
                'arrival': timestamp,
                'departure': timestamp, # Using arrival as departure for simplicity if same
                'headsign': trip['trip_headsign'],
                'direction': int(trip['direction_id']),
                'date': date_str,
                'stopId': stop_time['stop_id']
            })

# Sort
final_schedule.sort(key=lambda x: x['arrival'])

# Save
with open('src/data/static_schedule.json', 'w') as f:
    json.dump(final_schedule, f, indent=2)

print(f"✅ Generated {len(final_schedule)} scheduled stops")

File: scripts/gtfs/generate_static_json.ts
import fs from 'fs';
import path from 'path';
import { parse } from 'csv-parse/sync';
import { format, addDays } from 'date-fns';
import { parseParisTime } from '../../src/utils/date';

const GTFS_DIR = 'gtfs_data';

const TARGET_ROUTE_NAMES = new Set(['E1', '20']); // Support 20 as fallback
const TARGET_STOP_NAMES = ['GERZAT Champfleuri', 'Patural'];
const TARGET_MAIN_STOP = 'GERZAT Champfleuri';

function readCsv(fileName: string): any[] {
  const filePath = path.join(GTFS_DIR, fileName);
  if (!fs.existsSync(filePath)) return [];
  const fileContent = fs.readFileSync(filePath, 'utf-8');
  const content = fileContent.charCodeAt(0) === 0xFEFF ? fileContent.slice(1) : fileContent;
  return parse(content, { columns: true, skip_empty_lines: true });
}

function parseGtfsParisTime(dateStr: string, timeStr: string): number {
  const parts = timeStr.split(':').map(Number);
  const rawHours = parts[0] || 0;
  const minutes = parts[1] || 0;
  const seconds = parts[2] || 0;

  const extraDays = Math.floor(rawHours / 24);
  const normalizedHours = rawHours % 24;

  const midnightTs = parseParisTime(`${dateStr}T000000`);
  return midnightTs + extraDays * 86400 + normalizedHours * 3600 + minutes * 60 + seconds;
}

function main() {
  const now = new Date();
  const todayStr = format(now, 'yyyyMMdd');

  // Discover maximum date available across calendar.txt and calendar_dates.txt
  let maxCalendarDate = todayStr;
  for (const row of readCsv('calendar.txt')) {
    if (row.end_date && row.end_date > maxCalendarDate) {
      maxCalendarDate = row.end_date;
    }
  }
  for (const row of readCsv('calendar_dates.txt')) {
    if (row.date && row.date > maxCalendarDate) {
      maxCalendarDate = row.date;
    }
  }

  // Generate all dates from today up to maxCalendarDate (or at least 30 days ahead, max 90 days)
  const maxDays = 90;
  const dates: string[] = [];
  for (let i = 0; i < maxDays; i++) {
    const dStr = format(addDays(now, i), 'yyyyMMdd');
    if (dStr <= maxCalendarDate || i < 30) {
      dates.push(dStr);
    } else {
      break;
    }
  }

  console.log(`📅 Generating schedule for ${dates.length} days (${dates[0]} -> ${dates[dates.length - 1]})`);

  const targetRouteIds = new Set<string>();
  for (const row of readCsv('routes.txt')) {
    if (TARGET_ROUTE_NAMES.has(row.route_short_name)) {
      targetRouteIds.add(row.route_id);
      console.log(`✅ Found route: ${row.route_short_name} -> route_id=${row.route_id}`);
    }
  }

  if (targetRouteIds.size === 0) {
    console.log(`❌ Error: Could not find route with names ${Array.from(TARGET_ROUTE_NAMES).join(', ')}`);
    process.exit(1);
  }

  const gtfsConfig: Record<string, any> = {
    routeIds: Array.from(targetRouteIds),
    routeNames: Array.from(TARGET_ROUTE_NAMES),
    generatedAt: new Date().toISOString(),
  };

  fs.mkdirSync('src/data', { recursive: true });
  fs.writeFileSync('src/data/gtfs_config.json', JSON.stringify(gtfsConfig, null, 2) + '\n');
  console.log(`✅ Exported gtfs_config.json with route IDs: ${Array.from(targetRouteIds).join(', ')}`);

  const targetStopIds = new Set<string>();
  const mainStopIds = new Set<string>();
  const paturalStopIds = new Set<string>();

  for (const row of readCsv('stops.txt')) {
    const stopNameLower = row.stop_name.toLowerCase();
    for (const targetName of TARGET_STOP_NAMES) {
      if (stopNameLower.includes(targetName.toLowerCase())) {
        targetStopIds.add(row.stop_id);
        if (stopNameLower.includes(TARGET_MAIN_STOP.toLowerCase())) {
          mainStopIds.add(row.stop_id);
        }
        if (stopNameLower.includes('patural')) {
          paturalStopIds.add(row.stop_id);
        }
        console.log(`✅ Found stop: ${row.stop_name} -> stop_id=${row.stop_id}`);
      }
    }
  }

  if (targetStopIds.size === 0) {
    console.log(`❌ Error: Could not find any target stops`);
    process.exit(1);
  }

  gtfsConfig.stopIds = {
    all: Array.from(targetStopIds),
    champfleuri: Array.from(mainStopIds),
    patural: Array.from(paturalStopIds),
  };
  fs.writeFileSync('src/data/gtfs_config.json', JSON.stringify(gtfsConfig, null, 2) + '\n');
  console.log(`✅ Updated gtfs_config.json with Stop IDs`);

  const services: Record<string, any> = {};
  for (const row of readCsv('calendar.txt')) {
    services[row.service_id] = row;
  }

  const exceptions: Record<string, any[]> = {};
  for (const row of readCsv('calendar_dates.txt')) {
    if (!exceptions[row.date]) exceptions[row.date] = [];
    exceptions[row.date].push(row);
  }

  const trips: Record<string, any> = {};
  for (const row of readCsv('trips.txt')) {
    if (targetRouteIds.has(row.route_id)) {
      trips[row.trip_id] = row;
    }
  }
  console.log(`📋 Found ${Object.keys(trips).length} trips for target routes`);

  const tripStops = [];
  for (const row of readCsv('stop_times.txt')) {
    if (trips[row.trip_id] && targetStopIds.has(row.stop_id)) {
      tripStops.push(row);
    }
  }
  console.log(`📋 Found ${tripStops.length} stop times at target stops`);

  const finalSchedule: any[] = [];
  const DAYS = ['sunday', 'monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday'];

  for (const dateStr of dates) {
    if (dateStr.endsWith('0501')) continue;

    // Build noon time in UTC to avoid DST edge cases when computing day of week
    const dtUtc = new Date(Date.UTC(
      parseInt(dateStr.slice(0, 4)),
      parseInt(dateStr.slice(4, 6)) - 1,
      parseInt(dateStr.slice(6, 8)),
      12, 0, 0
    ));
    const dayOfWeek = dtUtc.getUTCDay();
    const daysKey = DAYS[dayOfWeek];

    const activeServices = new Set<string>();

    for (const [serviceId, service] of Object.entries(services)) {
      if (service.start_date <= dateStr && dateStr <= service.end_date) {
        if (service[daysKey] === '1') {
          activeServices.add(serviceId);
        }
      }
    }

    if (exceptions[dateStr]) {
      for (const exc of exceptions[dateStr]) {
        if (exc.exception_type === '1') {
          activeServices.add(exc.service_id);
        } else if (exc.exception_type === '2') {
          activeServices.delete(exc.service_id);
        }
      }
    }

    const tripEvents: Record<string, any> = {};

    for (const stopTime of tripStops) {
      const tripId = stopTime.trip_id;
      const stopId = stopTime.stop_id;

      if (tripEvents[tripId]) {
        const existingStop = tripEvents[tripId].stop_id;
        if (mainStopIds.has(stopId) && !mainStopIds.has(existingStop)) {
          tripEvents[tripId] = stopTime;
        }
      } else {
        if (!mainStopIds.has(stopId)) {
          const tripHeadsign = trips[tripId].trip_headsign.toUpperCase();
          if (!tripHeadsign.includes('BALLAINVILLIERS') && !tripHeadsign.includes('PATURAL')) {
            continue;
          }
        }
        tripEvents[tripId] = stopTime;
      }
    }

    for (const [tripId, stopTime] of Object.entries(tripEvents)) {
      const trip = trips[tripId];
      if (activeServices.has(trip.service_id)) {
        const timestamp = parseGtfsParisTime(dateStr, stopTime.arrival_time);

        finalSchedule.push({
          tripId,
          arrival: timestamp,
          departure: timestamp,
          headsign: trip.trip_headsign,
          direction: parseInt(trip.direction_id),
          date: dateStr,
          stopId: stopTime.stop_id
        });
      }
    }
  }

  finalSchedule.sort((a, b) => a.arrival - b.arrival);

  fs.writeFileSync('src/data/static_schedule.json', JSON.stringify(finalSchedule, null, 2) + '\n');
  console.log(`✅ Generated ${finalSchedule.length} scheduled stops`);
}

main();
